boot/fdehook/fdehook.go

```diff
@@ -0,0 +1,147 @@
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package fdehook
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"os"
+	"os/exec"
+	"path/filepath"
+
+	"github.com/snapcore/snapd/osutil"
+)
+
+func fdeHook(kernelDir string) string {
+	return filepath.Join(kernelDir, "meta/hooks/fde")
+}
+
+// Enabled returns whether the external FDE helper should be called
+func Enabled(kernelDir string) bool {
+	return osutil.FileExists(fdeHook(kernelDir))
+}
+
+// fdehookRuntimeMax is the maximum runtime a fdehook can execute
+func fdehookRuntimeMax() string {
+	// only used in tests (tests/main/fdehook)
+	if s := os.Getenv("DEBUG_FDEHOOK_RUNTIME_MAX"); s != "" {
+		return s
+	}
+	// XXX: reasonable default?
+	return "5m"
+}
+
+// fdeHookCmd returns a *exec.Cmd that runs the fdehook code with
+// (some) sandboxing around it. The sandbox will ensure that it's hard
+// for the hook to abuse that they are called in e.g. initrd.  But it
+// does not aim for perfect protection - the fdehooks are part of the
+// kernel snap/initrd so if someone wants to do mischief there are
+// easier ways by hacking initrd directly.
+func fdeHookCmd(kernelDir string, args ...string) *exec.Cmd {
+	cmd := exec.Command(
+		"systemd-run",
+		append([]string{
+			"--pipe", "--same-dir", "--wait", "--collect",
+			"--service-type=exec",
+			"--quiet",
+			// ensure we get some result from the hook
+			// within a reasonable timeout and output
+			// from systemd if things go wrong
+			fmt.Sprintf("--property=RuntimeMaxSec=%s", fdehookRuntimeMax()),
+			fmt.Sprintf(`--property=ExecStopPost=/bin/sh -c 'if [ "$EXIT_STATUS" != 0 ]; then echo "service result: $SERVICE_RESULT" 1>&2; fi'`),
+			// do not allow mounting, this ensures hooks
+			// in initrd can not mess around with
+			// ubuntu-data
+			"--property=SystemCallFilter=~@mount",
+			// add basic sandboxing to prevent messing
+			// around
+			// XXX: this maybe too strict, i.e. to unseal
+			//      the fdehook may need to write to some
+			//      crypto device in /dev which will not be
+			//      possible with "ProtectSystem=strict"
+			"--property=ProtectSystem=strict",
+			fdeHook(kernelDir),
+		}, args...)...)
+	return cmd
+}
+
+type UnlockParams struct {
+	SealedKey        []byte `json:"unsealed-key"`
+	VolumeName       string `json:"volume-name"`
+	SourceDevicePath string `json:"source-device-path"`
+	LockKeysOnFinish bool   `json:"lock-keys-on-finish"`
+}
+
+// Unlock unseals the key and unlocks the encrypted volume key specified
+// in params.
+//
+// This is usually called in the inird.
+func Unlock(kernelOrRootDir string, params *UnlockParams) (unsealedKey []byte, err error) {
+	jbuf, err := json.Marshal(params)
+	if err != nil {
+		return nil, err
+	}
+
+	cmd := fdeHookCmd(kernelOrRootDir, "--unlock")
+	cmd.Stdin = bytes.NewReader(jbuf)
+	// provide this via environment to make it easier for C based hooks
+	cmd.Env = append(os.Environ(),
```

Comment:
I tried this branch and ran the run-fdehook test, then I found `systemd-run` didn't pass these environment variables to the helper.
So I think we need to pass the env variables to `fdeHookCmd` and add additional arguments such as `--setenv=FDE_SEALED_KEY=...` there instead.

Reply:
Thanks for spotting this! I'm fixing this now and will improve the test